keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,149 @@
+    return tf.data.Dataset.from_tensor_slices(inputs).batch(batch_size or 32)
+
+
+def apply_dataset_fn(ds, fn):
+    return ds.map(fn, tf.data.AUTOTUNE).prefetch(tf.data.AUTOTUNE)
+
+
+class PipelineModel(keras.Model):
+    """A model which allows automatically applying preprocessing."""
+
+    def __init__(self, *args, include_preprocessing=True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.include_preprocessing = include_preprocessing
+
+    def preprocess_features(self, x):
+        """An overridable function which preprocesses features."""
+        return x
+
+    def preprocess_labels(self, y):
+        """An overridable function which preprocesses labels."""
+        return y
+
+    def preprocess(self, *args):
+        """An overridable function which preprocesses samples."""
+        x, y, sample_weight = keras.utils.unpack_x_y_sample_weight(args)
+        x, y = self.preprocess_features(x), self.preprocess_labels(y)
+        return keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+
+    # ========================================================================
+    # Below are overrides to keras.Model methods to apply the functions above.
+    # ========================================================================
+    def fit(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        validation_data=None,
+        validation_split=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess)
+
+        if validation_split:
+            raise ValueError(
+                "`validation_split` is not supported on this model. "
+                f"Received: `validation_split={validation_split}`."
+            )
+        if validation_data is not None:
+            val_tuple = keras.utils.unpack_x_y_sample_weight(validation_data)
+            validation_data = convert_inputs_to_dataset(*val_tuple, batch_size)
+            if self.include_preprocessing:
+                validation_data = apply_dataset_fn(
+                    validation_data, self.preprocess
+                )
+
+        return super().fit(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            validation_data=validation_data,
+            **kwargs,
+        )
+
+    def evaluate(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess)
+
+        return super().evaluate(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            **kwargs,
+        )
+
+    def predict(
```

Comment:
Some performance optimization? Converting to a dataset in the single sample case will be inefficient.

Add a check. If input is not a dataset and is short, avoid extra work here.

Reply:
I assume we could have a simple list comprehension or python `map` when the input length < x?

Reply:
Where `x = tf.data.AUTOTUNE` lol

Reply:
Yeah, we should check with @fchollet on this, he brought this up originally. I think there are some optimizations for this in core Keras we should mirror.